### PR TITLE
add informer

### PR DESCRIPTION
**Pull request by sylviah23**
#### What type of PR is this?
/kind feature
#### What is this PR about? / Why do we need it?
This adds an informer so that when a new node joins the cluster a controller will automatically patch the metadata label of that node to include volume/ENI count.

#### How was this change tested?
Unit test in labels_test.go and testing by manually adding new nodes to an EKS cluster

#### Does this PR introduce a user-facing change?
No

To test:
1) make cluster/create
2) make cluster/image
3) helm upgrade --install aws-ebs-csi-driver --namespace kube-system charts/aws-ebs-csi-driver --values driver_values.yaml --wait
4) make sylvia-test (runs E2E test)

Here driver_values.yaml looks something like:

```
image:
  pullPolicy: Always
  repository: 411511126293.dkr.ecr.us-west-2.amazonaws.com/aws-ebs-csi-driver
  tag: "4cdc804d8ba1b1f50e9c14cfce0ba284"
controller:
  logLevel: 4
  replicaCount: 1
  enableMetrics: true
node: 
  logLevel: 4
  enableMetrics: true
  metadataSources: Kubernetes
```

### CODE REVIEW

File: patch_nodes.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (AndrewSirenko) [inline patch_nodes.yaml:1]: Can we automatically install this for customers alongside the driver?

To do that you will have to add this file to `/charts/aws-ebs-csi-driver/templates/`. You can read up on [Helm](https://helm.sh/docs/chart_template_guide/getting_started/)) or see other PRs where we added new templates!

File: pkg/cloud/metadata/labels.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	return ""
+}
+
+func GetMetadata(client cloud.EC2API, region string, nodes *v1.NodeList) (map[string]enisVolumes, error) {
+	nodeIds := make([]string, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
+		nodeIds = append(nodeIds, parseNode(node.Spec.ProviderID))
+	}
+
+	resp, err := client.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{InstanceIds: nodeIds})
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:98]: If 100 nodes suddenly join the cluster, would we make 100 describeInstances calls? 

We might want to switch from EC2Client.DescribeInstaces to `cloud.getInstance` in cloud.go. That way the driver will batch up any getInstance function calls from the last .5 seconds into one big EC2 DescribeInstances call, which means that API is called less times in total.

You may need to export GetInstance

Another upside of this is that you'd get rid of the GetEC2 method you added to Cloud, which encourages developers to use the pre-made interfaces instead of re-inventing the wheel by calling EC2 APIs manually with the client. 

This may take some refactoring/time

File: cmd/main.go [hunk 3]
```diff
@@ -196,6 +200,19 @@ func main() {
 		klog.V(2).InfoS("Failed to setup k8s client", "err", err)
 	}
 
+	if options.Mode == driver.ControllerMode { // TODO: what to do in AllMode
```
Comment (AndrewSirenko) [inline cmd/main.go:203]: Today this only runs on controller startup and every time node is added to cluster, but is there a way of keeping node labels up to date? Perhaps by adding a goroutine that updates all labels ~hourly?

File: cmd/main.go [hunk 4]
```diff
@@ -144,6 +144,10 @@ func main() {
 	var md metadata.MetadataService
 	var metadataErr error
 
+	if options.Mode == driver.NodeMode {
+		time.Sleep(5e9) // TODO: for proof of concept remove later
+	}
```
Comment (AndrewSirenko) [inline cmd/main.go:149]: Longterm: In your next design doc you may want to consider whether we wait a few seconds on node pods for the additional labels to show up or not. 

In order to make that decision it will be useful to remove this sleep and time how long until labels show up.

To gather data for this decision, you can consider attempting to retry GetMetadata a few times before moving on, and timing how much time elapsed since node pod startup until it finally works

We can discuss offline

File: pkg/cloud/metadata/labels.go (hunk 2 above)
Comment (sylviah23) [reply pkg/cloud/metadata/labels.go:98]: Right now cloud.getInstance only takes in one instance as an argument. At cluster startup, is it okay to call cloud.getInstance for every node (since there is some sort of batching of API calls that is done as you described above), or do you think it is better to change cloud.getInstance to accept a list of instances as an argument?

Comment (AndrewSirenko) [reply pkg/cloud/metadata/labels.go:98]: IMO it's still cleaner to add a cloud.getInstances and use that just for controller startup case than leaking EC2Client outside of our cloud interface with GetEC2. So maybe also do that refactor in meantime?

File: cmd/main.go [hunk 5]
```diff
@@ -196,6 +200,34 @@ func main() {
...
+			updateMetadata()
+			for range ticker.C {
+				updateMetadata()
+			}
+		}()
+
+		informer := metadata.MetadataInformer(k8sClient, cloud)
+		stopCh := make(chan struct{})
+		informer.Start(stopCh)
+		informer.WaitForCacheSync(stopCh)
```
Comment (AndrewSirenko) [inline cmd/main.go:228]: IMO This logic is getting a bit complicated for main.go 

Can you wrap this in a function in another package?

File: cmd/main.go [hunk 6]
```diff
@@ -196,6 +200,34 @@ func main() {
...
+				return
+			}
+			err = metadata.UpdateMetadataEC2(k8sClient, cloud, nodes)
+			if err != nil {
+				klog.ErrorS(err, "unable to update ENI/Volume count on node labels")
+			}
+		}
+
+		go func() {
+			defer ticker.Stop()
```
Comment (AndrewSirenko) [inline cmd/main.go:218]: When does this defer get executed?

File: patch_nodes.yaml (hunk 1 above)
Comment (AndrewSirenko) [reply patch_nodes.yaml:1]: Now that you added the templates can you remove this file from commit?

File: pkg/cloud/cloud.go [hunk 7]
```diff
@@ -1490,7 +1490,8 @@ func describeInstances(ctx context.Context, svc EC2API, request *ec2.DescribeIns
 	return instances, nil
 }
 
-func (c *cloud) getInstance(ctx context.Context, nodeID string) (*types.Instance, error) {
+// GetInstance returns the instances of a single node.
```
Comment (AndrewSirenko) [inline pkg/cloud/cloud.go:1493]: nit: GetInstance returns the instance info associated with a single nodeID

File: pkg/cloud/metadata/labels_test.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+func TestMetadataInformer(t *testing.T) {
+	testCases := []struct {
+		name             string
+		newNode          *corev1.Node
+		expectedMetadata map[string]enisVolumes
+		expErr           error
+	}{
+		{
+			name: "success: normal",
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels_test.go:43]: Is this the only test case? 

E.g: What about the case where labels don't exist?

File: pkg/cloud/metadata/labels_test.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+func TestGetMetadata(t *testing.T) {
+	testCases := []struct {
+		name             string
+		instances        []*types.Instance
+		nodes            *corev1.NodeList
+		expectedMetadata map[string]enisVolumes
+		expErr           error
+	}{
+		{
+			name:      "success: normal",
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels_test.go:125]: Is this the only case we expect?

File: pkg/cloud/metadata/metadata.go [hunk 10]
```diff
@@ -90,9 +93,26 @@ func NewMetadataService(cfg MetadataServiceConfig, region string) (MetadataServi
...
 // We do not refresh blockDeviceMappings because IMDS only reports data from when instance starts (As of April 2025).
-func (m *Metadata) UpdateMetadata() error {
+func (m *Metadata) UpdateMetadata(kubeconfig string) error {
 	if m.IMDSClient == nil {
-		// IMDS not available, skip updates
+		k8sClient, err := DefaultKubernetesAPIClient(kubeconfig)()
+
+		if err != nil {
+			klog.V(2).InfoS("Failed to setup k8s client", "err", err)
+		}
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/metadata.go:102]: By the time UpdateMetadata is called, the driver already may have setup k8sClient. Can we reuse that k8sClient instead of recreating a new one? (similar to how this fcn reuses IMDSClient?)

File: pkg/cloud/metadata/k8s.go [hunk 11]
```diff
@@ -141,14 +142,49 @@ func KubernetesAPIInstanceInfo(clientset kubernetes.Interface) (*Metadata, error
...
+	} else {
+		klog.V(2).InfoS("num-volumes label not found on node, defaulting to 0 volumes")
+		volumes = 0
+	}
+	return volumes
+}
+
+func getENIs(node *corev1.Node) int {
+	var enis int
+	if val, ok := node.GetLabels()["num-ENIs"]; ok {
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/k8s.go:178]: We usually prefix our labels with `ebs.csi.aws.com/` so folks know where this label came from!

Could you do the same? 

Also consider 
- turning it into a const
- and make it all lowercase
- And switching to eni-count & `non-csi-ebs-volumes-count` or smth like that? 

See https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/65d50842b62bad5a9b57deef102b00e6b33541d6/pkg/driver/constants.go#L180

File: Makefile [hunk 12]
```diff
@@ -239,6 +239,12 @@ sub-push-a1compat:
 .PHONY: all-push
 all-push: sub-push sub-push-fips sub-push-a1compat
 
+# QUESTION: should my tests be run in the current CI jobs or a separate job
+
+.PHONY: sylvia-test
+sylvia-test:
+	cd tests/e2e && ginkgo -v -focus="Node labeling volumes and ENIs"
```
Comment (AndrewSirenko) [inline Makefile:246]: Can we prepend "[Disruptive]"  to the name of each of your new tests? 

And then instead you can ginkgo focus on "\[Disruptive\]"?

That way if we separate your tests into different CI job we can SKIP disruptive tests in old CI job.

Comment (AndrewSirenko) [reply Makefile:246]: Similar to our current `[single-az]` in name of other tests https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/a445e6f19b747479f452f62faec1967b58e63d3d/tests/e2e/nvme_metrics.go#L39

File: cmd/main.go [hunk 13]
```diff
@@ -196,6 +200,10 @@ func main() {
 		klog.V(2).InfoS("Failed to setup k8s client", "err", err)
 	}
 
+	if options.Mode == driver.ControllerMode {
+		metadata.ContinuousUpdateLabels(k8sClient, cloud, 60)
```
Comment (AndrewSirenko) [inline cmd/main.go:204]: This 60 is a magic number! https://en.wikipedia.org/wiki/Magic_number_(programming)#:~:text=In%20computer%20programming%2C%20a%20magic,replaced%20with%20a%20named%20constant.

Can we turn it into a named constant?

File: tests/sanity/fake_sanity_metadata_service.go [hunk 14]
```diff
@@ -34,7 +35,7 @@ func newFakeMetadataService(id string, r string, az string, oa arn.ARN) *fakeMet
 	}
 }
 
-func (m *fakeMetadataService) UpdateMetadata() error {
+func (m *fakeMetadataService) UpdateMetadata(kubernetes.Interface) error {
```
Comment (AndrewSirenko) [inline tests/sanity/fake_sanity_metadata_service.go:38]: High level comment:

Edge cases to worry about:
1. Double counting root volume 
2. Confirm we aren't double counting first ENI (I don't think we are! But worth double checking) 
3. Double counting CSI Managed volumes which Kubernetes already takes into consideration when scheduling pods. (Should not affect allocatable count) 

Consider changing e2e tests to check allocatable account against hardcoded,  [known EBS max volume attachment limits](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html)! So that we know if we are double counting.

File: pkg/cloud/metadata/labels.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+	if providerID != "" {
+		parts := strings.Split(providerID, "/")
+		if len(parts) > 0 {
+			return parts[len(parts)-1]
+		}
+	}
+	return ""
+}
+
+func getMetadata(client cloud.Cloud, nodes *v1.NodeList) (map[string]enisVolumes, error) {
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:128]: Can we have a small function comment explaining what (and why) the function does?

File: pkg/cloud/metadata/labels_test.go (hunk 8 above)
Comment (AndrewSirenko) [reply pkg/cloud/metadata/labels_test.go:44]: +1 to adding unhappy case.

File: pkg/cloud/metadata/metadata.go [hunk 16]
```diff
@@ -90,9 +94,20 @@ func NewMetadataService(cfg MetadataServiceConfig, region string) (MetadataServi
 
 // UpdateMetadata refreshes ENI information.
 // We do not refresh blockDeviceMappings because IMDS only reports data from when instance starts (As of April 2025).
-func (m *Metadata) UpdateMetadata() error {
+func (m *Metadata) UpdateMetadata(k8sClient kubernetes.Interface) error {
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/metadata.go:97]: On further thought, lets not worry about this UpdateMetadata feature for your intern project. 

Unless you clearly have extra time after your main PR gets merged. I think there are some dragons not worth thinking about for now here.

File: pkg/cloud/metadata/metadata.go [hunk 17]
```diff
@@ -90,9 +94,20 @@ func NewMetadataService(cfg MetadataServiceConfig, region string) (MetadataServi
 
 // UpdateMetadata refreshes ENI information.
 // We do not refresh blockDeviceMappings because IMDS only reports data from when instance starts (As of April 2025).
```
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply pkg/cloud/metadata/metadata.go:96]: Nvm, see next comment

File: charts/aws-ebs-csi-driver/templates/controller.yaml [hunk 18]
```diff
@@ -422,6 +422,40 @@ spec:
...
+            {{- . | toYaml | nindent 12 }}
+            {{- end }}
+          {{- with .Values.controller.envFrom }}
+          envFrom:
+            {{- . | toYaml | nindent 12 }}
+          {{- end }}
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /csi
+          {{- with default .Values.controller.resources .Values.sidecars.patcher.resources }}
```
Comment (AndrewSirenko) [inline charts/aws-ebs-csi-driver/templates/controller.yaml:451]: Do we need this? If not delete!

File: charts/aws-ebs-csi-driver/templates/controller.yaml [hunk 19]
```diff
@@ -422,6 +422,40 @@ spec:
...
+          resources:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
+          {{- with .Values.sidecars.patcher.securityContext }}
+          securityContext:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
```
Comment (AndrewSirenko) [inline charts/aws-ebs-csi-driver/templates/controller.yaml:458]: Do we foresee needing any special security context?

File: charts/aws-ebs-csi-driver/values.schema.json [hunk 20]
```diff
@@ -1091,6 +1091,63 @@
...
+              "default": [],
+              "items": {
+                "type": "string"
+              }
+            },
+            "resources": {
+              "type": ["object", "null"],
+              "default": null
+            },
+            "logLevel": {
```
Comment (AndrewSirenko) [inline charts/aws-ebs-csi-driver/values.schema.json:1110]: I don't think logLevel is currently used in templates/controller.yaml

Can you use it somehwere?

File: charts/aws-ebs-csi-driver/values.schema.json [hunk 21]
```diff
@@ -1091,6 +1091,63 @@
...
+                },
+                "tag": {
+                  "type": "string"
+                },
+                "repository": {
+                  "type": "string",
+                  "default": "public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver"
+                }
+              }
+            },
```
Comment (AndrewSirenko) [inline charts/aws-ebs-csi-driver/values.schema.json:1145]: are these still used anywhere?

File: charts/aws-ebs-csi-driver/values.schema.json [hunk 22]
```diff
@@ -1091,6 +1091,63 @@
...
+            "securityContext": {
```
Comment (AndrewSirenko) [inline charts/aws-ebs-csi-driver/values.schema.json:1146]: Make sure you remove anything you don't use from values.schema.json

File: charts/aws-ebs-csi-driver/templates/controller.yaml (hunk 19 above)
Comment (AndrewSirenko) [reply charts/aws-ebs-csi-driver/templates/controller.yaml:458]: Oh nvm if other containers have it maybe we need it...

File: pkg/cloud/metadata/k8s.go [hunk 23]
```diff
@@ -89,13 +92,13 @@ func DefaultKubernetesAPIClient(kubeconfig string) KubernetesAPIClient {
...
-			return nil, err
+			return nil, nil, err
 		}
-		return clientset, nil
+		return clientset, config, nil
 	}
 }
 
-func KubernetesAPIInstanceInfo(clientset kubernetes.Interface) (*Metadata, error) {
+func KubernetesAPIInstanceInfo(clientset kubernetes.Interface, sylvia bool) (*Metadata, error) {
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/k8s.go:101]: Yeah lets change every instance of sylvia to some placeholder name like "ec2Labels"?

File: pkg/cloud/metadata/k8s.go [hunk 24]
```diff
@@ -22,28 +22,31 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/util/cert"
 	"k8s.io/klog/v2"
 )
 
-type KubernetesAPIClient func() (kubernetes.Interface, error)
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/k8s.go:35]: What is this new `*rest.Config` used for..?

File: cmd/main.go [hunk 25]
```diff
@@ -191,11 +197,16 @@ func main() {
 		klog.FlushAndExit(klog.ExitFlushTimeout, 1)
 	}
 
-	k8sClient, err := cfg.K8sAPIClient()
+	k8sClient, k8sConfig, err := cfg.K8sAPIClient()
 	if err != nil {
 		klog.V(2).InfoS("Failed to setup k8s client", "err", err)
 	}
 
+	// TODO: Question: should I put this in the same case statement as where we check for prestop hook
```
Comment (AndrewSirenko) [inline cmd/main.go:205]: Yeah lets put this in the switch statement ASAP actually.

Worst case duplicate the other code that you need inside the switch statement itself

That way we can spot what is duplicated and can be refactored into helper function 

OR done before the switch statement (like k8sClient and logging)

File: docs/options.md [hunk 26]
```diff
@@ -20,4 +20,4 @@ There are a couple of driver options that can be passed as arguments when starti
...
-| metadata-sources                      | imds         | imds,kubernetes                                  | Dictates which sources are used to retrieve instance metadata. The driver will attempt to rely on each source in order until one succeeds. Valid options include 'imds' and 'kubernetes'.                                                                                                                                                                                                                                                    |
+| metadata-sources                      | imds         | imds,kubernetes,sylviakubernetes                                  | Dictates which sources are used to retrieve instance metadata. The driver will attempt to rely on each source in order until one succeeds. Valid options include 'imds' and 'kubernetes'.                                                                                                                                                                                                                                                    |
```
Comment (AndrewSirenko) [inline docs/options.md:23]: Need to add description of sylviakubernetes metadata source to /docs/install.md 

Can put placeholder text there for now

File: pkg/cloud/metadata/labels.go [hunk 27]
```diff
@@ -0,0 +1,322 @@
...
+		}
+	}
+	return ""
+}
+
+// getMetadata() calls the EC2 API to get the number of ENIs and non-CSI managed volumes attached to each node.
+func getMetadata(ctx context.Context, cloudclient cloud.Cloud, nodes *v1.NodeList, pvInformer cache.SharedIndexInformer) (map[string]enisVolumes, error) {
+	nodeIds := make([]string, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
+		nodeIds = append(nodeIds, parseNode(node.Spec.ProviderID))
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:254]: What if customer has a cluster that isn't all EC2 instances? (e.g. fargate nodes OR hybrid cluster)

Maybe we should only append to nodeIds if we are sure it is an ec2 instance? (We can either look for a known tag or that id begins with `i-`)

Otherwise I think cloudclient.GetInstances will keep failing with InstanceNotFound

File: pkg/cloud/metadata/k8s.go [hunk 28]
```diff
@@ -107,6 +110,35 @@ func KubernetesAPIInstanceInfo(clientset kubernetes.Interface) (*Metadata, error
...
+			if enis, volumes, err = getEC2ENIsVolumes(node); err != nil {
+				klog.ErrorS(err, "Get ENI and volume labels failed, retrying...")
+				//nolint: nilerr // Want to catch retry all errs until context times out
+				return false, nil
+			}
+			return true, nil
+		})
+
+		if backoffErr != nil {
+			klog.ErrorS(backoffErr, "Get ENI and volume labels failed after multiple retries, falling back to Kubernetes metadata")
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/k8s.go:137]: We don't necessarily know if we're falling back to kuberentes metadata. Can we just have the log message say "Getting node ENI and volume labels failed after multiple retries"

File: pkg/cloud/metadata/k8s.go [hunk 29]
```diff
@@ -146,9 +178,52 @@ func KubernetesAPIInstanceInfo(clientset kubernetes.Interface) (*Metadata, error
...
+	var enis int
+	if val, ok := node.GetLabels()[ENIsLabel]; ok {
+		var err error
+		enis, err = strconv.Atoi(val)
+		if err != nil {
+			klog.ErrorS(err, "failed to convert number of ENIs label to int")
+			return 1, err
+		}
+	} else {
+		klog.V(2).InfoS("num-ENIs label not found on node")
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/k8s.go:225]: You might want to use the full label name for num-enis. You can use the const you already defined in codebase. Same thing for the volume error message

File: pkg/cloud/metadata/labels.go [hunk 30]
```diff
@@ -0,0 +1,322 @@
...
+
+	l, err := rl.NewFromKubeconfig(
+		rl.LeasesResourceLock,
+		lockNamespace,
+		lockName,
+		rl.ResourceLockConfig{
+			Identity: instanceID, //TODO: question: what's a better identity to use
+		},
+		k8sConfig,
+		time.Second*10,
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:67]: Refactor this into a const so it is clear what this value is.

File: pkg/cloud/metadata/labels.go [hunk 31]
```diff
@@ -0,0 +1,322 @@
...
+		RenewDeadline: time.Second * 10,
+		RetryPeriod:   time.Second * 2,
+		Name:          lockName,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				klog.InfoS("This controller is the leader for updating metadata labels")
+				continuousUpdateLabels(ctx, clientset, cloud, updateTime)
+			},
+			OnStoppedLeading: func() { klog.InfoS("This controller is no longer the leader for updating metadata labels") },
+			OnNewLeader:      func(identity string) { klog.InfoS("the leader for updating metadata labels is", "", identity) },
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:84]: Is there a reason the first "The" is not capitilized?

File: pkg/cloud/metadata/labels.go [hunk 32]
```diff
@@ -0,0 +1,322 @@
...
+			if *vol.Ebs.VolumeId == pv {
+				nonCSIVolumes -= 1
+			}
+		}
+	}
+	return nonCSIVolumes
+}
+
+func isEBSVolume(pv *v1.PersistentVolume) bool {
+	if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "ebs.csi.aws.com" {
```
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply pkg/cloud/metadata/labels.go:144]: Add unit test for the migrated case

File: pkg/cloud/metadata/labels.go [hunk 33]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+func updateMetadataEC2(ctx context.Context, kubeclient kubernetes.Interface, cloudclient cloud.Cloud, nodes *v1.NodeList, pvInformer cache.SharedIndexInformer) error {
+	ENIsVolumeMap, err := getMetadata(ctx, cloudclient, nodes, pvInformer)
+	if err != nil {
+		klog.ErrorS(err, "unable to get ENI/Volume count")
+		return err
+	}
+
+	err = patchNodes(ctx, nodes, ENIsVolumeMap, kubeclient)
```
Comment (AndrewSirenko) [inline pkg/cloud/metadata/labels.go:232]: PR Wide comment: Can we get some happy case errors? 

Maybe at verbosity 4 log every time we successfully patch node and say what node it was, and what labels we just patched there.
